sklearn/tests/test_multiclass.py

```diff
@@ -668,6 +668,19 @@ def test_ovo_float_y():
 def test_ecoc_exceptions():
     ecoc = OutputCodeClassifier(LinearSVC(random_state=0))
     assert_raises(ValueError, ecoc.predict, [])
+    # test that base estimator raises the right error if it
+    # does not accept sparse matrices as input
+
+    class DummyNB(MultinomialNB):
```

Comment:
I am almost sure that we have already something that will do the job.
Could you look at `sklearn.utils._mocking.CheckingClassifier`. I assume that we can pass `check_array` with `accept_sparse=False`.

Also, you can create a new test just for  your case:

```python
def test_ecoc_delegate_sparse_base_estimator():
    # Non-regression test for
    # https://github.com/scikit-learn/scikit-learn/issues/17218
    ecoc = OutputCodeClassifier(CheckClassifier(...))
    with pytest.raises(TypeError, match=err_msg):
        ecoc.fit(...)

    # smoke test to check when sparse input should be supported
    ecoc = OutputCodeClassifier(RandomForestClassifier())
    ecoc.fit(...)
```

Reply:
Thanks for the suggestion I didnt know there was a class like `CheckingClassifier`. The problem with it is that `OutputCodecClassifier` performs an estimator check and requires the base estimator to implement `decision_function` or `predict_proba`. This means I'll still have to subclass CheckingClassifier to satisfy that requirement. Do you think I should still go ahead and use it that way or is there an alternative?

Reply:
Also this line 
https://github.com/scikit-learn/scikit-learn/blob/b0cd5955363630adaf9f9f0e5095fd65f5ad6f54/sklearn/utils/_mocking.py#L93
causes problems for sparse input. I set an error: `TypeError: sparse matrix length is ambiguous; use getnnz() or shape[0]`

Reply:
Uhm we should use `_num_samples`. I will make a PR to correct those things such that you can reuse the `CheckingClassifier` for these purposes.

Reply:
I submitted #17259. I still need a bit of work but it should be pretty fast.
Once this merge, we will be able to use `CheckingClassifier` for the job here.

Reply:
Just saw it now, thanks. I left a question about the possibility of implementing a default `predict_proba` or `decision_function` since it is required by some MetaEstimators like OutputCodeClassifier

Reply:
OK so #17259 was merged so we can build on it.

You can reuse something really similar to: https://github.com/scikit-learn/scikit-learn/pull/17259/files#diff-3e19e35c6e380d907095984bde6789f5R82-R91

Reply:
Thanks, I have just pushed the changes following the merge and your previous recommendation. I think this is ready for a review since my local tests pass.

Reply:
Also, not sure how I would best squash my commits into one after merging from master.